app/javascript/mastodon/service_worker/web_push_notifications.js

```diff
@@ -75,7 +87,7 @@ const handleNotificationClick = (event) => {
       }
     } else {
       event.notification.close();
-      resolve(self.clients.openWindow(event.notification.data.url));
+      resolve(openUrl(event.notification.data.url));
```

Comment:
I don't know how handleNotificationClick is attached exactly, but I remember reading that React events are synthetic and are reused between calls. So perhaps the thing @beatrix-bitrot is experiencing is event.notification.data.url referring to something from an earlier call. You could try making a copy of the data you're interested in outside of the Promise.